SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,157 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.hubspot.baragon.models.BaragonRequestState;
+import com.hubspot.baragon.models.UpstreamInfo;
+import com.hubspot.singularity.LoadBalancerRequestType;
+import com.hubspot.singularity.LoadBalancerRequestType.LoadBalancerRequestId;
+import com.hubspot.singularity.SingularityDeploy;
+import com.hubspot.singularity.SingularityLoadBalancerUpdate;
+import com.hubspot.singularity.SingularityRequest;
+import com.hubspot.singularity.SingularityRequestWithState;
+import com.hubspot.singularity.SingularityTask;
+import com.hubspot.singularity.SingularityTaskId;
+import com.hubspot.singularity.SingularityTaskIdsByStatus;
+import com.hubspot.singularity.data.DeployManager;
+import com.hubspot.singularity.data.RequestManager;
+import com.hubspot.singularity.data.TaskManager;
+import com.hubspot.singularity.helpers.RequestHelper;
+import com.hubspot.singularity.hooks.LoadBalancerClientImpl;
+import com.hubspot.singularity.mesos.SingularitySchedulerLock;
+
+
+@Singleton
+public class SingularityUpstreamChecker {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SingularityUpstreamChecker.class);
+  private final LoadBalancerClientImpl lbClient;
+  private final TaskManager taskManager;
+  private final RequestManager requestManager;
+  private final DeployManager deployManager;
+  private final RequestHelper requestHelper;
+  private final SingularitySchedulerLock lock;
+
+  @Inject
+  public SingularityUpstreamChecker(LoadBalancerClientImpl lbClient,
+                                    TaskManager taskManager,
+                                    RequestManager requestManager,
+                                    DeployManager deployManager,
+                                    RequestHelper requestHelper,
+                                    SingularitySchedulerLock lock) {
+    this.lbClient = lbClient;
+    this.taskManager = taskManager;
+    this.requestManager = requestManager;
+    this.deployManager = deployManager;
+    this.requestHelper = requestHelper;
+    this.lock = lock;
+  }
+
+  private List<SingularityTask> getActiveTasksForRequest(String requestId) {
+    final Optional<SingularityTaskIdsByStatus> taskIdsByStatusForRequest = requestHelper.getTaskIdsByStatusForRequest(requestId);
+    if (taskIdsByStatusForRequest.isPresent()) {
+      final List<SingularityTaskId> activeHealthyTaskIdsForRequest = taskIdsByStatusForRequest.get().getHealthy();
+      final Map<SingularityTaskId, SingularityTask> activeTasksForRequest = taskManager.getTasks(activeHealthyTaskIdsForRequest);
+      return new ArrayList<>(activeTasksForRequest.values());
+    }
+    return new ArrayList<>();
```

Comment:
I think this is another case where I'd lean towards throwing an exception and essentially skipping to the next request we need to check. If the taskIds is absent that means we couldn't get task data for some reason. I'd be wary of making any sort of LB update when we did not correctly fetch task data.